### PR TITLE
Fix ice when error reporting recursion errors

### LINKED ISSUE #90319: regression: ICE ErrorReporting Overflow should not reach report_selection_err call
* https://crater-reports.s3.amazonaws.com/beta-1.57-1/beta-2021-10-22/reg/syn-next-1.0.0-rc4/log.txt

```
[INFO] [stdout] error: internal compiler error: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:846:17: ErrorReporting Overflow should not reach `report_selection_err` call
```

This crate already failed to [build on 1.56](https://crater-reports.s3.amazonaws.com/beta-1.57-1/1.56.0/reg/syn-next-1.0.0-rc4/log.txt), but it seems worth trying to fix the ICE in case it affects other code as well.

### PR DESCRIPTION
**Pull request by light4**
Fixes: #90319, #92148, #93955

### CODE REVIEW

**Comment by light4** [comment]
Note there is another PR #91238 fixes the ICE in a different way.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 1]
```diff
@@ -924,8 +925,11 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 self.tcx.sess.delay_span_bug(span, "`ErrorReported` without an error");
                 return;
             }
-
-            Overflow => {
+            // Already reported.
+            Overflow(OverflowError::Error(ErrorReported)) => {
+                return;
```
[the remark this answers is not in the document]
Comment (light4) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:930]: Change this back to a `bug!` will cause ICE.

**Comment by tom7980** [comment]
@estebank I think this fix is cleaner than mine, see my comment in the changed files but I'd go with this over #91238 as mine introduces a bunch of overhead for this specific issue that we don't really need.

**Comment by Dylan-DPC** [comment]
closing this based on comment by author https://github.com/rust-lang/rust/pull/94391#issuecomment-1051947057

**Comment by light4** [comment]
@Dylan-DPC should not we close #91238 instead of this?

**Comment by Dylan-DPC** [comment]
ah yes :D

**Comment by estebank** [comment]
Apologies for the delay in reviewing. Can you check if adding the `delay_span_bug` call before the return causes any issues? Having that call protects us from every accidentally compiling if other parts of the codebase change in the future. Either way, once we know the answer to that, r=me.

**Comment by estebank** [comment]
Tepidly nominating for backport, as this fixes a stable regression that has existed for multiple stable releases.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (light4) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:935]: updated

**Comment by apiraino** [comment]
Beta backport declined, discussion of compiler team [on Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-03-31.20.2354818.html#277287220).

@estebank as mentioned in the discussion, please feel free to advocate otherwise

@rustbot label -beta-nominated

**Review by Hamidb14 — APPROVED** [review]

**Pull request merged**